File: 20-valid-parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, input_str):
        """
        :type s: str
        :rtype: bool
        """
        valid = {
        "{": "}", 
        "[": "]", 
        "(": ")"
        }
        stack = []
        for char in input_str:
            if char in valid.keys():
                stack.append(char)
            elif char in valid.values():
                if not stack:
                    return False
                last_open = stack.pop()
                if valid[last_open] != char:
                    return False
            else:
                return False
        return len(stack) == 0
        

File: 20-valid-parentheses/valid-parentheses.go
func isValid(s string) bool {
    stack := []string{}
    pairs := map[string]string{"(": ")", "{": "}", "[": "]"}

    for _, c := range s {
        if _, ok := pairs[string(c)]; ok {
            stack = append(stack, string(c))
        } else {
            if len(stack) == 0 || pairs[stack[len(stack)-1]] != string(c) {
                return false
            }
            stack = stack[:len(stack)-1]
        }
    }

    return len(stack) == 0 
}